Rewrote testFatePrintAndSummaryCommandsWithInProgressTxns for elasticity

kevinrr888 opened the pull request:
closes #4614

Rewrote the test `testFatePrintAndSummaryCommandsWithInProgressTxns` for elasticity which was merged in from https://github.com/apache/accumulo/pull/4631 to 2.1 and main. This is essentially the same for the test of META transactions as it is in 2.1/main. However, for USER transactions (which are stored in a table not in ZK like in 2.1/main), I simulate a transaction returning an UNKNOWN status in `AdminUtil.getTransactionStatus()` (which is what would occur if the transaction has completed since aquiring the list of transactions) and ensure this transaction is included in the returned value and does not cause any errors.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/FateOpsCommandsIT.java:
```diff
@@ -592,6 +610,124 @@ protected void testFateDeleteCommand(FateStore<TestEnv> store, ServerContext sct
+
+    ReadOnlyFateStore.ReadOnlyFateTxStore<TestEnv> mockedFateTxStore1, mockedFateTxStore2,
+        mockedFateTxStore3;
+    if (isUserStore) {
```
Opened #4688 after looking at this test because it highlighted that the behavior of the two fate stores is different.  If get #4688 is merged then this test could possibly be simplified alot because it can assume a single behavior instead of accommodating two.

kevinrr888 replied:
yes, that is a nice change

kevinrr888 commented:
From https://github.com/apache/accumulo/pull/4688, `MetaFateStore.FateTxStoreImpl.top()` was changed to no longer throw a NNE and instead return null. I checked to see if any of the other methods called in the try-catch block of `getTransactionStatus()` could throw a NNE and they do not. `top()` was the only method that did not handle and threw the NNE if encountered. So, the original issue of a NNE exception occurring in `getTransactionStatus()` causing the print/summary commands to fail is no longer a problem in elasticity (from what I can tell). So, now the check for a NNE exception in `getTransactionStatus()` is unneccessary but doesn't hurt to keep in case of future changes to `MetaFateStore.FateTxStoreImpl` resulting in these methods throwing a NNE. For `testFatePrintAndSummaryCommandsWithInProgressTxns()`, I mock that the NNE can occur in the try-catch block (even though this won't happen in the current implementation) and ensure that it is properly handled.

keith-turner commented:
From a tech debt perspective it may be best to remove the handling of NoNodeException from AdminUtil since its not expected to happen now.  Having the handling for NoNodeException may cause someone to spend time carrying it forward when refactoring AdminUtil, which would be uneeded work.  If we remove the handling for NNE, not sure there is anything to test.  We have the test for the fate stores to verify they handle the non-existent case in a uniform way.  It would probably be good to put those expectations in the javadoc on the fate store interfaces, but that is not something for this PR.  Wondering if this PR should be abandoned or changed to remove the NNE handling.

keith-turner commented:
Maybe there is something to test for this changes still if NNE handling is removed.  We want to test that admin util handles the case of something showing up in list and then no longer existing when its accessed.  Just want to ensure that admin util code does not fail in that case or report junk.

kevinrr888 commented:
Yeah, I wasn't sure how I would test it after removal of the NNE handling in AdminUtil and wasn't sure if this handling would be better left in or removed. I have just removed the NNE handling in AdminUtil and reworked the test. Now, the test ensures AdminUtil.getTransactionStatus() functions as expected if there is a FateId returned from the list() command that doesn't exist. This simplified the test a lot too.
